### DISCUSSION TITLE
(very) high CPU usage of concourse-web after submitting (many) tasks

### DISCUSSION THREAD

**Discussion by x3nb63** [concourse/concourse#7138] [General]
I happened to have a need to batch process a number of things ... so, given Concourse is  a "thing doer" I just did:

```
fly -t my-cluster execute --config task.yaml
```

about 10.000 times in a row

which results in this CPU load on the `concourse-web` container:
```
    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND       
1277253 root      20   0 3135480   2.3g  49312 R  2122   1.8   3213:55 concourse
```
Which is about 90% CPU utilization of that server - guess that hits its hardware limit (as it does other things as well). So the system operates in this condition for about 30 minutes and appears stable, while going through the tasks.

The probably unusual specialty of my cluster setup is  that each worker only does one job at a time, achieved by a placement strategy of limit-active-tasks=1.
Given there are 10 workers, only 10 tasks should run concurrently.

So, I would expect that my 10.000 submissions require quite some wallclock-time, while the hardware should stay pretty much idle. That is both, the nodes as well as the server hardware.

How come that `concourse-web` part - which I assume contains the task scheduler - burns so many CPU cycles?

My total container count is at ~1250 relatively constant since my mass-submission: thats probably some relation to existing pipelines and the 10 workers. The pipelines, by the way, sit idle, no builds going on.

**Comment by x3nb63** [comment]
> My total container count is at ~1250 relatively constant since my mass-submission: thats probably some relation to existing pipelines and the 10 workers. The pipelines, by the way, sit idle, no builds going on.

just notice: all these containers are of type `check` (looking with  `fly containers --print-table-headers`)

is there no type `task`?

**Reply by taylorsilva** [reply]
Most containers are usually check containers. Task containers will disappear once they're done running unless they failed, then they stick around 5 days or until the next build of the job they were part of. Not sure how that works for one-off tasks done with fly execute.

Back to your high CPU thing though. What's probably happening is that the web nodes are all running the web-side components for running each task. You have configured your concourse to not overload the workers with tasks, but the logic that ensures that rate-limiting is enforced happens at an individual container scheduling level. So your 10,000 tasks result in 10,000+ go routines on the web node, with each one constantly checking  "is there a worker available for me to run this container on?"

If you looked at each build from the web UI (or streamed the logs) you would see a long stream of "waiting for worker..." over and over again. This is being outputted by a go routine on the web node, which of course takes some CPU each time  it checks. Multiply that by 10,000 tasks and I'm guessing that's how you end up doing nothing while using 90% of the CPU.

Of course there's a lot that's not ideal about this situation and we can all brainstorm many solutions for reducing this CPU usage. But this is where the code is at right now given past design decisions and the refactors that we have been able to make so far. I hope that answers your question :)

**Reply by x3nb63** [reply]
yes, answers it. :-)

... its a curiosity question, cause I have some HPC background and so didn't really think about 10,000 batch-type things will cause much. Its not how Concourse is used typically, I see now.

I don't plan to do that often. ;-)

**Comment by x3nb63** [comment]
hmm, guess I still managed to deadlock something?!

It spend the night with this high CPU load but did not appear to execute any of my tasks. ... and I don't know how to stop that?

 `fly` appears not of help, cause I cant distinguish actual checks from the tasks, nor stop anything.

So I am counting things in the database and see this:

```
concourse=> select count(*) from containers;
 count 
-------
   390
(1 row)

concourse=> select count(*) from jobs;
 count 
-------
   131
(1 row)

concourse=> select count(*) from build_events;
 count  
--------
 501752
(1 row)
```

the `build_events` vary: generally increasing but sometimes decreasing by 10 .. 50, as if something got done.

I don't understand enough of the DB structure to know what I am looking at / what to clean.


EDIT: 
- also updated from 7.2 to 7.3.1 just now - but it keeps (re)loading the CPU after starting
- triggering regular build jobs in pipelines appears unaffected - they execute as usual

**Reply by x3nb63** [reply]
here I go, something achieves this:

```
concourse=> SELECT name,active_tasks FROM workers ORDER BY name;
        name        | active_tasks 
--------------------+--------------
 concourse-worker-0 |           60
 concourse-worker-1 |           68
 concourse-worker-2 |           69
 concourse-worker-3 |           68
 concourse-worker-4 |           67
 concourse-worker-5 |           70
 concourse-worker-6 |           93
 concourse-worker-7 |           96
 concourse-worker-8 |           66
 concourse-worker-9 |           72
 cw01               |            0
 cw02               |            0
(12 rows)
```

remember, my system has `limit_active_tasks=1` so its very effectively stuck

**Reply by taylorsilva** [reply]
If you look at the builds table you can get the id for each build which should represent one of the pending builds. With those ids you can then abort them which should kill the related go routine on the web node `fly abort-build --build`
